### PR TITLE
Implemented additions in OER code generator

### PR DESCRIPTION
**Pull request by Futsch1**
The next PR... One is to come after this one ;-)

I have created a larger change by implementing extension additions for the OER C code generator.

Now, the generator will encode and decode data taking additions into account. The major change in the generator was to add the interpretation of the upper bit of the presence mask and to encode/decode the additions if required. 

To encode additions, the generated length of a type is required in the OER data. This required an additional functionality in the code generator: Calculating the encoded type length. For some types, this is possible during code generation time and adding a fixed value. There are some types where the length depends on the values they have during run time. Therefore, the calculated encoded type length is a list of ints or strings, depending if the value is fixed (int) or variable (string). Each choice requires an additional helper function where its length is calculated during runtime.

Decoding is easier, but additions that are not yet known in the decoder have to be skipped using the decoded length.

I have added tests for both the cases that data from the past and data from the future is read (in terms of ASN.1 specification).

The UPER code generator is now also aware of the extension addition bit, but will behave differently (since I did not implement additions there): If an addition is either in the data to be encoded or if the decoded data contains additions, an error is returned. So, the new test sequence AE can be encoded and decoded, the others (AF and AG) not.

While implementing and testing, I came across two bugs I fixed along with my implementation, if you prefer, I can also create separate pull requests for them:
- OER additions were encoded/decoded wrongly in the codes/oer.py module: the length octet contains the number of bytes for the addition mask, not the number of additions.
- NULL was encoded wrongly in the additions of OER and PER: The addition presence bit for the NULL field was never set, even if the field was indicated in the data. However, the field must be encoded with a length of zero.
There are regression tests added for both bugs in test_per.py and test_oer.py.

### CODE REVIEW

File: tests/test_per.py [hunk 1]
```diff
@@ -153,7 +158,9 @@ def test_integer(self):
             ('R',               4294967296, b'\x80\x01\x00\x00\x00\x00'),
             ('S',
              {'a': True, 'b': 0, 'c': True},
-             b'\x90\x27\x10\x80')
+             b'\x90\x27\x10\x80'),
+            ('T', {'a': True, 'b': None},
+             b'\xc0\x40\x00')
```
[the remark this answers is not in the document]
Comment (Futsch1) [reply tests/test_per.py:163]: I trust that tool, too, and also used it for my implementation. In this special case, when decoding the output of the asn1tools/the tested bytestring c04000 in the playground, it gives the following:
`ASN1STEP: Decoding PDU #1 :

rec1value T ::= 
{
  --TYPE INFORMATION: SEQUENCE
  --OFFSET: 0,0
  --extension flag: <.1>
  a TRUE,
    --TYPE INFORMATION: BOOLEAN
    --OFFSET: 0,1; LENGTH: 0,1
    --contents: <1>
  --extension preamble size flag: <0> (preamble size <= 64)
  --extension preamble length: <00000.0> (decoded as 1)
  --extension preamble: <1>
    --bit #0 = 1: extension 'b' is present
  --padding: <000000>
  --extension length: .00 (decoded as 0)
  b NULL
    --TYPE INFORMATION: NULL
    --OFFSET: 3,0; LENGTH: 0,0
}
--TOTAL LENGTH: 3,0
Successfully decoded 3 bytes.
rec1value T ::= 
{
  a TRUE,
  b NULL
}`

I think this is a corner case of the specification, where the question is how NULL shall be encoded as an open type. My interpretation would be that NULL can be encoded either as a length 1 value of 0 or as a length 0 value.

[the remark this answers is not in the document]
Comment (eerimoq) [reply tests/test_per.py:163]: Ok, sounds fair. Smaller encoding is better :P

File: tests/test_per.py [hunk 2]
```diff
@@ -91,6 +91,11 @@ def test_integer(self):
             "    b INTEGER (-10000..704000000000000001), "
             "    c BOOLEAN "
             "} "
+            "T ::= SEQUENCE { "
+            "  a BOOLEAN, "
+            "  ..., "
+            "  b NULL "
+            "} "
```
Comment (eerimoq) [inline tests/test_per.py:98]: The `T` sequence should be moved to another test as it does not test integers. Move it to the sequence test.

Comment (Futsch1) [reply tests/test_per.py]: Oh, you're right, I overlooked that. Is changed now.

File: tests/test_per.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ def test_integer(self):
             "    a BOOLEAN, "
             "    b INTEGER (-10000..704000000000000001), "
             "    c BOOLEAN "
-            "} "
+            "} "
```
Comment (eerimoq) [inline tests/test_per.py:93]: Remove trailing whitespaces.

**Comment by eerimoq** [comment]
Ok, so I havn't done a proper code review, but the the parts I've looked at looks good, and there are some new tests, so I'll merge.

**Pull request merged**